Create view to display translation inconsistencies

flodolo opened the pull request:
I think this should be ready for a review.

One note: I didn't use "perfect search" for the search link because it fails quite often, for example with HTML fragments coming from DTD with leading/trailing new lines.

Also added a "splash screen" to avoid burning resources (and let people wait for seconds) when randomly open the menu.

Fixes #410

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+/**
+ * Consistency class
+ *
+ * Functions used to analyze consistency
```
I would precise "to analyse translation consistency".

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+         $ignore_string = function ($key, $value) use ($repository_type) {
+             // Determine if a string should be excluded.
+
+             // Ignore strings of lenght 1 (e.g. accesskeys) and empty strings.
```
length

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+         }
+
+         $ignore_string = function ($key, $value) use ($repository_type) {
+             // Determine if a string should be excluded.
```
put the comment before the anonymous function to make it readable with folding on

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+             }
+
+             // Exclude CSS rules, "widht:" or "height:"
+             if (strpos($value, 'width:') !== false || strpos($value, 'height:') !== false) {
```
We have a Strings::inString() method, I think it would be interesting to extend it to support arrays like that:
```php
/**
 * Check if $needles are in $haystack
 *
 * @param  string  $haystack String to analyse
 * @param  array   $needles  The strings to look for
 * @return boolean True if the $haystack string contains all strings in $needles
 */
public static function inString($haystack, $needles)
{
    foreach ((array) $needles as $needle) {
        if (mb_strpos($haystack, $needle, $offset = 0, 'UTF-8') === false) {
            return false;
        }
    }

    return true;
}
```
Then you could replace the above check with:
```php
if (Strings::inString($value, ['width:', 'height:'])) {
    return true;
}
```
I could not measure any performance hit in this case because of the use of mb_strpos vs strpos

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+                 if (strpos($key, 'accessibility.properties') !== false) {
+                     return true;
+                 }
+             } elseif ($repository_type == 'desktop') {
```
you probably don't need an elseif and can just use an if

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+
+             if ($repository_type == 'gaia') {
+                 // Ignore plural forms containing [] in the key
+                 if (strpos($key, '[') !== false && strpos($key, '[') !== false) {
```
Did you mean to search for [ and ]? I think there is a typo since you are looking for [ twice. Also, would be another candidate for `` if (Strings::inString($value, ['[', ']']))```

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+                 */
+                 if (strpos($key, 'AccessFu.properties') !== false) {
+                     return true;
+                 }
```
Again, I think this would be an improvement with a common comment above
```php
if (Strings::inString($key, ['region.properties', 'defines.inc', 'AccessFu.properties'])) {
    return true;
}
```

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+
+/*
+    Since this is a very heavy view for Aurora (default repo), we display
+    an introductory text when reaching the page and use a paramter to
```
parameter

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+    an introductory text when reaching the page and use a paramter to
+    determine if we actually want to display the strings.
+*/
+$display_results = isset($_GET['display']) ? true : false;
```
isset() returns a boolean, you can just write:
```php
$display_results = isset($_GET['display']);
```

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+$display_results = isset($_GET['display']) ? true : false;
+$reference_locale = Project::getReferenceLocale($repo);
+
+if ($display_results) {
```
actually, since you use $display_results only once, you could just as well put
```php
if (isset($_GET['display'])) {
```

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+
+if ($display_results) {
+    // Get non-empty strings for reference and locale in this repository
+    $source_strings = array_filter(Utils::getRepoStrings($reference_locale, $repo), 'strlen');
```
You probably don't need to use a callback function like strlen because by default array_filter() filters results evaluating to false and an empty string evaluates to false.

pascalchevrel commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,121 @@
+                 return true;
+             }
+
+             // Exclude CSS widht values like '38em'
```
width

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+
+    if (! empty($source_strings)) {
+        $duplicated_strings_english = Consistency::findDuplicates($source_strings);
+        $duplicated_strings_translation = Consistency::findDuplicates($target_strings);
```
You are using this temporary array only once line 37. If you do line 37 like that:
```php
    // Get the strings duplicated in English but not in the localization
    $missing_duplicates = array_diff_key(
        $duplicated_strings_english, 
        Consistency::findDuplicates($target_strings)
    );
```
You save 2MB of memory

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+    } else {
+        $strings_number = 0;
```
The else clause adds complexity, especially since there are many ifs before and I had to look closely to see which if it was linked to. I think that since this is the default value, it should be set up early like line 23 with a comment indicating how it will be used in the view.

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+    $source_strings = array_filter(Utils::getRepoStrings($reference_locale, $repo), 'strlen');
+    $target_strings = array_filter(Utils::getRepoStrings($locale, $repo), 'strlen');
+
+    if (! empty($source_strings)) {
```
I would test for true and return early, this way you gain an indentation level on all the code that follows and that makes it easier to read: 
```php
// We are going to show the placeholder text and not do any calculation
if (! $display_results) {
    return;
}

// Get non-empty strings for reference and locale in this repository
$source_strings = array_filter(Utils::getRepoStrings($reference_locale, $repo));
$target_strings = array_filter(Utils::getRepoStrings($locale, $repo));
…
```

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+        $missing_duplicates = array_diff_key($duplicated_strings_english, $duplicated_strings_translation);
+    }
+
+    if (! empty($duplicated_strings_english)) {
```
I would again return early and gain an indentation level on the following code:
```php

// No duplicated strings to analyze, get back to the view.
if (empty($duplicated_strings_english)) {
    return;
}
```

flodolo commented:
Thanks, I think I've addressed all review comments. I went with a slightly different approach for the inString method, adding the possibility to match any or all needles.

flodolo commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,115 @@
+        foreach ($strings_array as $key => $value) {
+            if (strcasecmp($previous_value, $value) === 0) {
+                $duplicates[$previous_key] = $previous_value;
+                $duplicates[$key] = $strings_array[$key];
```
This should simply be
```PHP
$duplicates[$key] = $value;
```

flodolo commented on app/models/consistency.php:
```diff
@@ -0,0 +1,85 @@
+    $missing_duplicates = array_diff_key(
+        $duplicated_strings_english,
+        Consistency::findDuplicates($target_strings)
+    );
```
I was trying to understand why a string supposed to be duplicated would only display one result in the translation column, and realized that the logic is flawed and I need to find a completely different approach.

Example: there are two strings for 'closed' in English, but one is translated as 'chiusa', the other one as 'chiuso', so they're not duplicated in Italian.

flodolo commented on app/models/consistency.php:
```diff
@@ -0,0 +1,85 @@
+    $duplicated_strings_english = Consistency::findDuplicates($source_strings);
+
+    // Filter English duplicates before comparing them with the localization
+    $duplicated_strings_english = Consistency::filterStrings($duplicated_strings_english, $repo);
```
Should filter strings before searching for duplicates. For example "closed" is available in two strings, but one needs to be filtered out, so it's not actually duplicated.

flodolo commented on app/classes/Transvision/Consistency.php:
```diff
@@ -0,0 +1,115 @@
+             }
+
+             // Exclude CSS rules, "width:" or "height:"
+             if (Strings::inString($key, ['width:', 'height:'])) {
```
`$value`, not `$key`

pascalchevrel commented on app/models/consistency.php:
```diff
@@ -0,0 +1,79 @@
+if ($display_results) {
+    // Get non-empty strings for reference and locale in this repository
+    $source_strings = array_filter(Utils::getRepoStrings($reference_locale, $repo), 'strlen');
+    $target_strings = array_filter(Utils::getRepoStrings($locale, $repo), 'strlen');
```
Thinking about it, array_filter should be done in findDuplicates() and not in the model, that way you get it defined only once and you don't need to think that the data provided has to be filtered for empty strings if the method is used in another view

flodolo commented:
Completely changed the model's logic, I think it's sound now.

I filter out know problematic strings and then search for duplicated in English (was doing the opposite before). Then I traverse the array checking translations for each of these strings, and keep the ones with more than one translation.

flodolo commented:
Looking into the failing test

pascalchevrel commented:
The new version is much faster, I think that the splash screen is no longer useful if the view has no performance hit.

flodolo commented:
True, getting rid of it.

pascalchevrel commented on app/classes/Transvision/Strings.php:
```diff
@@ -64,15 +64,36 @@ public static function endsWith($haystack, $needles)
+            return true;
+        }
+
+        return false;
```
That seems a bit complicated, also $total_needles isn't used. I would rather write that:
```php
    public static function inString($haystack, $needles, $match_all = false)
    {
        $match = 0;
        foreach ((array) $needles as $needle) {
            if (mb_strpos($haystack, $needle, $offset = 0, 'UTF-8') !== false) {
                // If I need to match any needle, I can stop at the first found
                if (! $match_all) {
                    return true;
                }
                $match++;
            }
        }

        return $match == count($needles);
    }
```

flodolo replied:
Agreed, definitely more readable.
